# ✈️ Flight Price Prediction

## Notebook 01: Data Understanding

### 🎯 Objective
The goal of this notebook is to understand the dataset before performing any preprocessing or model building.

In this notebook, we will:

- Load the training and testing datasets
- Explore the structure of the data
- Understand the features
- Identify missing values
- Check duplicate records
- Classify features into numerical and categorical
- Record initial observations



In [1]:
# Suppress warning messages
import warnings
warnings.filterwarnings("ignore")

# Data manipulation
import pandas as pd
import numpy as np

# Visualization
import matplotlib.pyplot as plt
import plotly.express as px

# Display all columns
pd.set_option("display.max_columns", None)

In [2]:
# Load training dataset
train_df = pd.read_excel("../data/raw/Train.xlsx")

# Load testing dataset
test_df = pd.read_excel("../data/raw/Test.xlsx")

In [3]:
# Display first five rows
train_df.head()

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info,Price
0,IndiGo,24/03/2019,Banglore,New Delhi,BLR → DEL,22:20,01:10 22 Mar,2h 50m,non-stop,No info,3897
1,Air India,1/05/2019,Kolkata,Banglore,CCU → IXR → BBI → BLR,05:50,13:15,7h 25m,2 stops,No info,7662
2,Jet Airways,9/06/2019,Delhi,Cochin,DEL → LKO → BOM → COK,09:25,04:25 10 Jun,19h,2 stops,No info,13882
3,IndiGo,12/05/2019,Kolkata,Banglore,CCU → NAG → BLR,18:05,23:30,5h 25m,1 stop,No info,6218
4,IndiGo,01/03/2019,Banglore,New Delhi,BLR → NAG → DEL,16:50,21:35,4h 45m,1 stop,No info,13302


In [4]:
test_df.head()

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info
0,Jet Airways,6/06/2019,Delhi,Cochin,DEL → BOM → COK,17:30,04:25 07 Jun,10h 55m,1 stop,No info
1,IndiGo,12/05/2019,Kolkata,Banglore,CCU → MAA → BLR,06:20,10:20,4h,1 stop,No info
2,Jet Airways,21/05/2019,Delhi,Cochin,DEL → BOM → COK,19:15,19:00 22 May,23h 45m,1 stop,In-flight meal not included
3,Multiple carriers,21/05/2019,Delhi,Cochin,DEL → BOM → COK,08:00,21:00,13h,1 stop,No info
4,Air Asia,24/06/2019,Banglore,Delhi,BLR → DEL,23:55,02:45 25 Jun,2h 50m,non-stop,No info


In [5]:
print("Training Dataset Shape :", train_df.shape)
print("Testing Dataset Shape  :", test_df.shape)

Training Dataset Shape : (10683, 11)
Testing Dataset Shape  : (2671, 10)


In [6]:
print("Training Dataset Columns\n")
print(train_df.columns.tolist())

Training Dataset Columns

['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route', 'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops', 'Additional_Info', 'Price']


In [7]:
train_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   Airline          10683 non-null  str  
 1   Date_of_Journey  10683 non-null  str  
 2   Source           10683 non-null  str  
 3   Destination      10683 non-null  str  
 4   Route            10682 non-null  str  
 5   Dep_Time         10683 non-null  str  
 6   Arrival_Time     10683 non-null  str  
 7   Duration         10683 non-null  str  
 8   Total_Stops      10682 non-null  str  
 9   Additional_Info  10683 non-null  str  
 10  Price            10683 non-null  int64
dtypes: int64(1), str(10)
memory usage: 1.8 MB


In [8]:
train_df.describe()

,Price
count,10683.000000
mean,9087.064121
std,4611.359167
min,1759.000000
25%,5277.000000
50%,8372.000000
75%,12373.000000
max,79512.000000


In [9]:
train_df.describe(include="object")

,Airline,Date_of_Journey,Source,Destination,Route,Dep_Time,Arrival_Time,Duration,Total_Stops,Additional_Info
count,10683,10683,10683,10683,10682,10683,10683,10683,10682,10683
unique,12,44,5,6,128,222,1343,368,5,10
top,Jet Airways,18/05/2019,Delhi,Cochin,DEL → BOM → COK,18:55,19:00,2h 50m,1 stop,No info
freq,3849,504,4537,4537,2376,233,423,550,5625,8345


In [10]:
missing_values = train_df.isnull().sum()

missing_values

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [11]:
missing_percentage = (
    train_df.isnull().sum() / len(train_df)
) * 100

missing_percentage

Airline            0.000000
Date_of_Journey    0.000000
Source             0.000000
Destination        0.000000
Route              0.009361
Dep_Time           0.000000
Arrival_Time       0.000000
Duration           0.000000
Total_Stops        0.009361
Additional_Info    0.000000
Price              0.000000
dtype: float64

In [12]:
duplicates = train_df.duplicated().sum()

print("Duplicate Rows :", duplicates)

Duplicate Rows : 220


In [13]:
train_df.dtypes

Airline              str
Date_of_Journey      str
Source               str
Destination          str
Route                str
Dep_Time             str
Arrival_Time         str
Duration             str
Total_Stops          str
Additional_Info      str
Price              int64
dtype: object

In [14]:
categorical_columns = train_df.select_dtypes(
    include="object"
).columns.tolist()

print(categorical_columns)

['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route', 'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops', 'Additional_Info']


In [15]:
numerical_columns = train_df.select_dtypes(
    exclude="object"
).columns.tolist()

print(numerical_columns)

['Price']


In [16]:
for column in train_df.columns:
    print(f"{column:<20} : {train_df[column].nunique()}")

Airline              : 12
Date_of_Journey      : 44
Source               : 5
Destination          : 6
Route                : 128
Dep_Time             : 222
Arrival_Time         : 1343
Duration             : 368
Total_Stops          : 5
Additional_Info      : 10
Price                : 1870


In [17]:
target = "Price"

print("Target Variable :", target)

Target Variable : Price


In [18]:
X = train_df.drop(columns="Price")

y = train_df["Price"]

print("Feature Shape :", X.shape)
print("Target Shape  :", y.shape)

Feature Shape : (10683, 10)
Target Shape  : (10683,)


# Initial Observations

## Dataset Summary

- Training dataset contains **10,683 rows**.
- Testing dataset contains **2,671 rows**.
- Target variable is **Price**.
- Most features are categorical.
- Date and time columns are currently stored as strings.
- Route and Total_Stops contain missing values.
- Duration is stored as text and will require preprocessing.
- Feature engineering will be necessary before model training.